### PR TITLE
feat(oriyomi): Create The Notification Collection and add hooks to ideaCard and Comment Collection

### PR DESCRIPTION
**Pull request by oriyomibadmus**

### CODE REVIEW

File: src/api/comment/content-types/comment/lifecycles.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module.exports = {
+
+    async afterCreate(event) {
+
+        const ideaCard = event.result.idea_card
+        const ideaName = event.result.idea_card.ideaName
+        const commentAuthor = event.result.author
+        const commentText = event.result.text
+        const timeCreated = event.result.createdAt
```
Comment (Enjoy2Live) [inline src/api/comment/content-types/comment/lifecycles.js:9]: maybe if we use object destructing here it'd be better

File: src/api/comment/content-types/comment/lifecycles.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+        await strapi.entityService.create('api::notification.notification', {
+            data: {
+              Title: commentAuthor + " commented on " + ideaName,
+              Content: commentText,
+              Tag: "New Comment",
+              idea_card: ideaCard.id,
+              TimeCreated: timeCreated,
+              Read: false,
```
Comment (Enjoy2Live) [inline src/api/comment/content-types/comment/lifecycles.js:18]: how does the state changes from unread to read? are we supposed to send some sort of request to backend?

File: src/api/idea-card/content-types/idea-card/lifecycles.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+module.exports = {
+
+    async afterCreate(event) {
+
+        const ideaCard = event.result.id
+        const ideaName = event.result.ideaName
+        const timeCreated = event.result.createdAt
```
Comment (Enjoy2Live) [inline src/api/idea-card/content-types/idea-card/lifecycles.js:7]: maybe if we use object destructing here it'd be better

**Review by Enjoy2Live — CHANGES_REQUESTED** [review]
That's promising, thanks for implementing such a needed feature @oriyomibadmus .
I appropriate the simplicity of the PR and how everything is documenting itself.

I left some comments here to get better idea with some requested changes

File: src/api/notification/content-types/notification/schema.json [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      "relation": "oneToOne",
+      "target": "api::idea-card.idea-card"
+    },
+    "Content": {
+      "type": "string"
+    },
+    "Read": {
+      "type": "boolean"
+    },
+    "Tag": {
```
Comment (chungthuang) [inline src/api/notification/content-types/notification/schema.json:29]: What does tag mean here?

File: src/api/notification/content-types/notification/schema.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  },
+  "options": {
+    "draftAndPublish": false
+  },
+  "pluginOptions": {},
+  "attributes": {
+    "Title": {
+      "type": "string"
+    },
+    "idea_card": {
```
Comment (chungthuang) [inline src/api/notification/content-types/notification/schema.json:18]: Would we need to create a field for each new collection that wants notifications? Maybe we can create an enum field for the collection type (idea-card, comment, etc) and another field for ID of the object that triggered the notification.
The downside is we don't have a relationship at the database level, so we can't do things like cascade delete directly from database.

File: src/api/comment/content-types/comment/lifecycles.js (hunk 2 above)
Comment (Enjoy2Live) [inline src/api/comment/content-types/comment/lifecycles.js:18]: how can we bind the notification to its appropriate user?
for example only idea_author and people interested in the idea would want to get notified but not everyone else

File: src/api/notification/content-types/notification/schema.json
Comment (chungthuang) [inline src/api/notification/content-types/notification/schema.json:1]: Should we store the ID of the user that triggered the notification? For example, the user that commented on an idea.

Comment (chungthuang) [reply src/api/notification/content-types/notification/schema.json:1]: Actually maybe we don't need it. We can get the author from the comment and author from the idea card.

File: src/api/comment/content-types/comment/lifecycles.js (hunk 1 above)
Comment (chungthuang) [reply src/api/comment/content-types/comment/lifecycles.js:9]: Like `const {ideaCard, timeCreated} = event.result`.

File: src/api/notification/content-types/notification/schema.json (hunk 4 above)
Comment (oriyomibadmus) [reply src/api/notification/content-types/notification/schema.json:29]: There are plans in the future to sort all notifications into categories. Thought adding a tag field might be helpful

Comment (chungthuang) [reply src/api/notification/content-types/notification/schema.json:29]: Maybe that's not needed if we do https://github.com/dev-launchers/strapiv4/pull/103/files#r1412896943.

File: src/api/comment/content-types/comment/lifecycles.js (hunk 2 above)
Comment (chungthuang) [reply src/api/comment/content-types/comment/lifecycles.js:18]: I think the requirement from product is notification will only be sent to 1 user. In the case of comment, only the idea owner will be notified. In the case of applicant, only the owner of the opportunity will be notified. Too many notifications can cause fatigue and people will start ignoring them.

File: src/api/comment/content-types/comment/lifecycles.js (hunk 1 above)
Comment (oriyomibadmus) [reply src/api/comment/content-types/comment/lifecycles.js]: @Enjoy2Live You are right! Done it!

File: src/api/idea-card/content-types/idea-card/lifecycles.js (hunk 3 above)
Comment (oriyomibadmus) [reply src/api/idea-card/content-types/idea-card/lifecycles.js]: Yes, Done

File: src/api/notification/content-types/notification/schema.json (hunk 5 above)
Comment (oriyomibadmus) [reply src/api/notification/content-types/notification/schema.json]: @chungthuang Just made the changes to the notification collection schema to reflect these

**Review by chungthuang — APPROVED** [review]

File: src/api/comment/content-types/comment/lifecycles.js (hunk 2 above)
Comment (Enjoy2Live) [reply src/api/comment/content-types/comment/lifecycles.js]: > I think the requirement from product is notification will only be sent to 1 user. In the case of comment, only the idea owner will be notified. In the case of applicant, only the owner of the opportunity will be notified. Too many notifications can cause fatigue and people will start ignoring them.

good point but my question is about the viability of changing the current solution we want to include multiple recipients of the same notification? (for example in the chat section someone might tag two people in the same comment)

Comment (Enjoy2Live) [reply src/api/comment/content-types/comment/lifecycles.js]: My guess is it's going to be an attribute in the notification collection but correct me if I'm wrong @oriyomibadmus .

btw happy new year! :smile:

**Review by Enjoy2Live — APPROVED** [review]
Awesome, that's a huge step.
Thank you guys

File: src/api/idea-card/content-types/idea-card/lifecycles.js [hunk 6]
```diff
@@ -1,9 +1,30 @@
...
+        Content: "You created idea " + ideaName,
+        Collection: "IdeaCard",
+        ObjectID: id,
+        TimeCreated: timeCreated,
+        Read: false
+      },
+    });
+
+  }
+}
```
Comment (chungthuang) [inline src/api/idea-card/content-types/idea-card/lifecycles.js:20]: I think you have an extra `,` after the rebase.

File: src/api/idea-card/content-types/idea-card/lifecycles.js [hunk 7]
```diff
@@ -1,9 +1,30 @@
...
+        Title:"A new idea has been created",
+        Content: "You created idea " + ideaName,
+        Collection: "IdeaCard",
+        ObjectID: id,
+        TimeCreated: timeCreated,
+        Read: false
+      },
+    });
+
+  }
```
Comment (chungthuang) [inline src/api/idea-card/content-types/idea-card/lifecycles.js:19]: To export multiple functions, you will need an `,` here.

**Pull request merged**